autoendpoint/src/server/routers/fcm/client.rs

```diff
@@ -0,0 +1,301 @@
+                ))
+                .expect("Project ID is not URL-safe"),
+            auth,
+            http,
+        })
+    }
+
+    /// Send the message data to FCM
+    pub async fn send(
+        &self,
+        data: HashMap<&'static str, String>,
+        token: String,
+        ttl: usize,
+    ) -> Result<(), FcmError> {
+        // Build the FCM message
+        let message = serde_json::json!({
+            "message": {
+                "token": token,
+                "android": {
+                    "ttl": format!("{}s", ttl),
+                    "data": data
+                }
+            }
+        });
+        let access_token = self.auth.token(OAUTH_SCOPES).await?;
+
+        // Make the request
+        let response = self
+            .http
+            .post(self.endpoint.clone())
+            .header("Authorization", format!("Bearer {}", access_token.as_str()))
+            .header("Content-Type", "application/json; UTF-8")
+            .body(message.to_string())
+            .timeout(TIMEOUT)
+            .send()
+            .await
+            .map_err(|e| {
+                if e.is_timeout() {
+                    FcmError::FcmRequestTimeout
+                } else {
+                    FcmError::FcmConnect(e)
+                }
+            })?;
+
+        // Handle error
+        if let Err(e) = response.error_for_status_ref() {
+            let data: FcmResponse = response
+                .json()
+                .await
+                .map_err(FcmError::DeserializeResponse)?;
+
+            return Err(match (e.status(), data.error) {
+                (Some(StatusCode::UNAUTHORIZED), _) => FcmError::FcmAuthentication,
+                (Some(StatusCode::NOT_FOUND), _) => FcmError::FcmNotFound,
+                (_, Some(error)) => FcmError::FcmUpstream {
+                    status: error.status,
+                    message: error.message,
+                },
+                _ => FcmError::FcmUnknown,
+            });
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Deserialize)]
+struct FcmResponse {
+    error: Option<FcmErrorResponse>,
+}
+
+#[derive(Deserialize)]
+struct FcmErrorResponse {
+    status: String,
+    message: String,
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::server::routers::fcm::client::FcmClient;
+    use crate::server::routers::fcm::error::FcmError;
+    use crate::server::routers::fcm::settings::FcmCredential;
+    use std::collections::HashMap;
+    use std::io::Write;
+    use std::path::PathBuf;
+    use tempfile::NamedTempFile;
+    use url::Url;
+
+    pub const PROJECT_ID: &str = "yup-test-243420";
+    const ACCESS_TOKEN: &str = "ya29.c.ElouBywiys0LyNaZoLPJcp1Fdi2KjFMxzvYKLXkTdvM-rDfqKlvEq6PiMhGoGHx97t5FAvz3eb_ahdwlBjSStxHtDVQB4ZPRJQ_EOi-iS7PnayahU2S9Jp8S6rk";
+
+    /// Write service data to a temporary file
+    pub fn make_service_file() -> NamedTempFile {
+        // Taken from the yup-oauth2 tests
+        let contents = serde_json::json!({
```

Comment:
kinda also wondering if we should put this as a json file to avoid some of the potential key flagging we're seeing. We could have the CI system include the key info as part of the build.

Reply:
The JSON file would still be flagged by the tool, since it's a secret that's checked into Git. Hiding it in CI would break local testing, unless everyone had a copy of the config (which causes issues for new people when first cloning and testing the repo).

Telling GitGuardian to ignore the file seems to be the most correct fix.